### PR TITLE
Restructure project layout

### PR DESCRIPTION
**Pull request by dio**
This patch introduces a new project structure to mimic the envoy project.
Hence making adding custom multi-filters to envoy binary easier, to match the description here: 

https://www.envoyproxy.io/docs/envoy/latest/intro/arch_overview/http_filters.html

Signed-off-by: Dhi Aurrahman <dio@rockybars.com>

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,7 +1,7 @@
 # Envoy filter example
 
 This project demonstrates the linking of additional filters with the Envoy binary.
-A new filter `echo2` is introduced, identical modulo renaming to the existing
+A new filter `my.net.filter.reader.example` is introduced, identical modulo renaming to the existing
```
Comment (htuch) [inline README.md]: Can you use reverse DNS, e.g. `net.some.filter.reader.example? This is the preferred convention.

File: README.md [hunk 2]
```diff
@@ -12,25 +12,30 @@ For an example of additional HTTP filters, see [here](http-filter-example).
...
 
-`bazel test @envoy//test/...`
+```
+$ bazel test @envoy//test/...
+```
 
 ## How it works
 
-The [Envoy repository](https://github.com/envoyproxy/envoy/) is provided as a submodule.
+The [Envoy repository](https://github.com/envoyproxy/envoy/) will be downloaded as a .zip via bazel's `http_archive`.
```
Comment (htuch) [inline README.md]: Yeah, this is an interesting change. Originally we used submodules to mimic what Lyft do when extending Envoy, but I think what you have here is the Bazel-first approach, so seems reasonable.

File: source/common/filter/reader/example/BUILD [hunk 3]
```diff
@@ -0,0 +1,43 @@
+package(default_visibility = ["//visibility:public"])
```
Comment (htuch) [inline source/common/filter/reader/example/BUILD]: FYI, we're considering reworking how filters are organized in the main Envoy repo. See https://github.com/envoyproxy/envoy/issues/2069. Do we want to get some informal buy-in to the new layout in that issue now and have this reflect that?

File: test/common/filter/reader/example/filter_integration_test.cc [hunk 4]
```diff
@@ -2,16 +2,16 @@
 #include "test/integration/utility.h"
 
 namespace Envoy {
-class Echo2IntegrationTest : public BaseIntegrationTest,
-                             public testing::TestWithParam<Network::Address::IpVersion> {
+class ExampleReaderIntegrationTest : public BaseIntegrationTest,
```
Comment (htuch) [inline test/common/filter/reader/example/filter_integration_test.cc]: Do you think it's worth reorganizing the HTTP filter as well in this repo-wide restructure?

File: source/common/filter/reader/example/BUILD (hunk 3 above)
Comment (dio) [reply source/common/filter/reader/example/BUILD]: Seems reasonable, let me try to reflect that in this example.

Comment (htuch) [reply source/common/filter/reader/example/BUILD]: You might want to comment on the issue that you're doing this to allow those who have reservations to speak now or forever hold their peace :)

File: README.md (hunk 1 above)
Comment (dio) [reply README.md]: Got it. So basically I have my own script (written in js using yeoman, not sure if the community like it haha. Apparently I need to work on my python skill to have it in `./tools` dir) something like:

```
$ ./generate net.reader --package net.some.filter.reader.example
$ ./generate http.encoder --package http.some.filter.encoder.example
```

to scaffold things out.

File: test/common/filter/reader/example/filter_integration_test.cc (hunk 4 above)
Comment (dio) [reply test/common/filter/reader/example/filter_integration_test.cc]: Yes, my intention actually to collect examples to make it ready for answering issues. For example something like https://github.com/envoyproxy/envoy/issues/2635.

Comment (dio) [reply test/common/filter/reader/example/filter_integration_test.cc]: I will put the http ones in this PR.

Comment (dio) [reply test/common/filter/reader/example/filter_integration_test.cc]: @htuch the current layout:

```
|-- source
|   |-- BUILD
|   `-- extensions
|       `-- filters
|           |-- http
|           |   `-- example
|           |       |-- README.md
|           |       |-- decoder
|           |       |   |-- BUILD
|           |       |   |-- README.md
|           |       |   |-- filter.cc
|           |       |   |-- filter.h
|           |       |   |-- filter.proto
|           |       |   `-- filter_config.cc
|           |       |-- decoder_encoder
|           |       |   |-- BUILD
|           |       |   |-- README.md
|           |       |   |-- filter.cc
|           |       |   |-- filter.h
|           |       |   |-- filter.proto
|           |       |   `-- filter_config.cc
|           |       `-- encoder
|           |           |-- BUILD
|           |           |-- README.md
|           |           |-- filter.cc
|           |           |-- filter.h
|           |           |-- filter.proto
|           |           `-- filter_config.cc
|           `-- network
|               `-- example
|                   |-- README.md
|                   `-- reader
|                       |-- BUILD
|                       |-- README.md
|                       |-- filter.cc
|                       |-- filter.h
|                       |-- filter.proto
|                       `-- filter_config.cc
|-- test
|   |-- BUILD
|   |-- envoy_binary_test.sh
|   |-- extensions
|   |   `-- filters
|   |       |-- http
|   |       |   `-- example
|   |       |       |-- decoder
|   |       |       |   `-- filter_test.cc
|   |       |       |-- decoder_encoder
|   |       |       |   `-- filter_test.cc
|   |       |       `-- encoder
|   |       |           `-- filter_test.cc
|   |       `-- network
|   |           `-- example
|   |               `-- reader
|   |                   `-- filter_test.cc
|   |-- fixtures
|   |   `-- configs
|   |       |-- BUILD
|   |       `-- v2
|   |           `-- echo.yaml
|   `-- integration
|       `-- extensions
|           `-- filters
|               |-- http
|               |   |-- BUILD
|               |   |-- example_decoder_encoder_test.cc
|               |   |-- example_decoder_test.cc
|               |   `-- example_encoder_test.cc
|               `-- network
|                   |-- BUILD
|                   `-- example_reader_test.cc
```

**Comment by htuch** [comment]
@dio this is great, let's use your layout as a strawman to discuss how tests should be structured in https://github.com/envoyproxy/envoy/issues/2069.

**Comment by dio** [comment]
Since this https://github.com/envoyproxy/envoy/pull/2894 is merged. Will close this and propose a new one.

**Pull request closed without merging**